components/payments/cmd/connectors/internal/storage/balances.go

```diff
@@ -2,59 +2,119 @@ package storage
 
 import (
 	"context"
+	"database/sql"
+	"errors"
 
 	"github.com/formancehq/payments/internal/models"
+	"github.com/uptrace/bun"
 )
 
 func (s *Storage) InsertBalances(ctx context.Context, balances []*models.Balance, checkIfAccountExists bool) error {
 	if len(balances) == 0 {
 		return nil
 	}
 
-	query := s.db.NewInsert().
-		Model((*models.Balance)(nil)).
-		With("cte1", s.db.NewValues(&balances)).
-		Column(
-			"created_at",
-			"account_id",
-			"balance",
-			"currency",
-			"last_updated_at",
-		)
-	if checkIfAccountExists {
-		query = query.TableExpr(`
-		(SELECT *
-			FROM cte1
-			WHERE EXISTS (SELECT 1 FROM accounts.account WHERE id = cte1.account_id)
-				AND cte1.balance != COALESCE((SELECT balance FROM accounts.balances WHERE account_id = cte1.account_id AND last_updated_at < cte1.last_updated_at AND currency = cte1.currency ORDER BY last_updated_at DESC LIMIT 1), cte1.balance+1)
-		) data`)
-	} else {
-		query = query.TableExpr(`
-		(SELECT *
-			FROM cte1
-			WHERE cte1.balance != COALESCE((SELECT balance FROM accounts.balances WHERE account_id = cte1.account_id AND last_updated_at < cte1.last_updated_at AND currency = cte1.currency ORDER BY last_updated_at DESC LIMIT 1), cte1.balance+1)
-		) data`)
+	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = tx.Rollback()
```

Comment:
Rolling back outside of a transaction (eg. in the case that `tx.Commit()` was successful) results in `noop` and not an error, so you shouldn't have to discard this error. Let's log it perhaps?

Reply:
Done thanks